parlai/core/utils.py

```diff
@@ -713,6 +715,52 @@ def __contains__(self, key):
         """Determine if text contains any offensive words in the filter."""
         return self.contains_offensive_language(key)
 
+    def str_segment(self, text, dict_agent):
+        freqs = dict_agent.freqs()
+
+        # Total number of word tokens
+        N = sum(freqs.values())
+
+        # Number of distinct words in the Vocab
+        V = len(freqs)
+
+        @lru_cache(maxsize=None)
+        def segment(text):
+            # Return a list of words that is the best segmentation of text.
+            if not text:
+                return []
+            candidates = ([first]+segment(rem) for first, rem in splits(text))
+            return max(candidates, key=prob_words)
+
+        def splits(text, L=20):
+            # Returns a list of all possible first and remainder tuples where
+            candidates = []
+            for i in range(min(len(text), L)):
+                candidates.append((text[:i+1], text[i+1:]))
+            return candidates
+
+        def prob_words(words):
+            # Returns probability for a sequence of words
+            return add(prob(w) for w in words)
+
+        def add(nums):
+            # Return the product of a sequence of numbers
+            return reduce(operator.add, nums, 1)
+
+        def avoid_long_words(word, N):
+            return 10./(N * 10**len(word))
+
+        # def laplace_smoothing(word, N, V):
+        #     return math.log((dict_agent.freq[word] + 1)/(N + V))
+
+        def prob(word):
+            if word in freqs:
+                return math.log(freqs[word]/N)
```

Comment:
already you should consider precomputing logN and returning math.log(freqs[word]) - logN.